L1Trigger/DTTriggerPhase2/plugins/DTTrigPhase2Prod.cc

```diff
@@ -0,0 +1,719 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "Geometry/Records/interface/MuonGeometryRecord.h"
+#include "Geometry/DTGeometry/interface/DTGeometry.h"
+#include "Geometry/DTGeometry/interface/DTLayer.h"
+
+#include "L1Trigger/DTTriggerPhase2/interface/MuonPath.h"
+#include "L1Trigger/DTTriggerPhase2/interface/constants.h"
+
+#include "L1Trigger/DTTriggerPhase2/interface/MotherGrouping.h"
+#include "L1Trigger/DTTriggerPhase2/interface/InitialGrouping.h"
+#include "L1Trigger/DTTriggerPhase2/interface/HoughGrouping.h"
+#include "L1Trigger/DTTriggerPhase2/interface/PseudoBayesGrouping.h"
+#include "L1Trigger/DTTriggerPhase2/interface/MuonPathAnalyzer.h"
+#include "L1Trigger/DTTriggerPhase2/interface/MuonPathAnalyzerPerSL.h"
+#include "L1Trigger/DTTriggerPhase2/interface/MuonPathAnalyzerInChamber.h"
+#include "L1Trigger/DTTriggerPhase2/interface/MuonPathAssociator.h"
+#include "L1Trigger/DTTriggerPhase2/interface/MPFilter.h"
+#include "L1Trigger/DTTriggerPhase2/interface/MPQualityEnhancerFilter.h"
+#include "L1Trigger/DTTriggerPhase2/interface/MPRedundantFilter.h"
+
+#include "DataFormats/MuonDetId/interface/DTChamberId.h"
+#include "DataFormats/MuonDetId/interface/DTSuperLayerId.h"
+#include "DataFormats/MuonDetId/interface/DTLayerId.h"
+#include "DataFormats/MuonDetId/interface/DTWireId.h"
+#include "DataFormats/DTDigi/interface/DTDigiCollection.h"
+#include "DataFormats/L1DTTrackFinder/interface/L1Phase2MuDTPhContainer.h"
+#include "DataFormats/L1DTTrackFinder/interface/L1Phase2MuDTPhDigi.h"
+
+// DT trigger GeomUtils
+#include "DQM/DTMonitorModule/interface/DTTrigGeomUtils.h"
+
+//RPC TP
+#include "DataFormats/RPCRecHit/interface/RPCRecHitCollection.h"
+#include <DataFormats/MuonDetId/interface/RPCDetId.h>
+#include "Geometry/RPCGeometry/interface/RPCGeometry.h"
+#include "L1Trigger/DTTriggerPhase2/interface/RPCIntegrator.h"
+
+#include <fstream>
+#include <iostream>
+#include <queue>
+#include <cmath>
+
+class DTTrigPhase2Prod : public edm::stream::EDProducer<> {
+  typedef std::map<DTChamberId, DTDigiCollection, std::less<DTChamberId>> DTDigiMap;
+  typedef DTDigiMap::iterator DTDigiMap_iterator;
+  typedef DTDigiMap::const_iterator DTDigiMap_const_iterator;
+
+  enum algo { Standard = 0, PseudoBayes = 1, HoughTrans = 2 };
+
+  enum scenario { MC = 0, DATA = 1, SLICE_TEST = 2 };
+
+public:
+  //! Constructor
+  DTTrigPhase2Prod(const edm::ParameterSet& pset);
+
+  //! Destructor
+  ~DTTrigPhase2Prod() override;
+
+  //! Create Trigger Units before starting event processing
+  void beginRun(edm::Run const& iRun, const edm::EventSetup& iEventSetup) override;
+
+  //! Producer: process every event and generates trigger data
+  void produce(edm::Event& iEvent, const edm::EventSetup& iEventSetup) override;
+
+  //! endRun: finish things
+  void endRun(edm::Run const& iRun, const edm::EventSetup& iEventSetup) override;
+
+  // Methods
+  int rango(metaPrimitive mp);
```

Comment:
since metaPrimitive is a heavy-weight object, it should always be passed by (const) ref
(in every function here and elsewhere)